Add DockerSecrets

HaoK opened the pull request:
Good job with the prototype @glennc pretty much what you did with some slight tweaking.

I think we need the lambda pattern as the basic building block for sources, adding overloads for every setting is meh.

aka `builder.AddXyz<Action<XyzSource> configureSource>()` should be available for all sources.

cc @divega

Meir017 commented on src/Microsoft.Extensions.Configuration.DockerSecrets/DockerSecretsConfigurationProvider.cs:
```diff
@@ -0,0 +1,71 @@
+                }
+                else
+                {
+                    throw new FileNotFoundException("Secrets directory doesn't exist and is not optional.");
```
message should include the word `Docker` to avoid confusing with `Microsoft.Extensions.Configuration.UserSecrets`. `Docker Secrets` instead of `Secrets`

Meir017 commented on src/Microsoft.Extensions.Configuration.DockerSecrets/DockerSecretsConfigurationProvider.cs:
```diff
@@ -0,0 +1,71 @@
+            var secretsDir = Source.FileProvider.GetDirectoryContents("/");
+            if (!secretsDir.Exists && !Source.Optional)
+            {
+                throw new FileNotFoundException("Secrets directory doesn't exist and is not optional.");
```
message should include the word `Docker` to avoid confusing with `Microsoft.Extensions.Configuration.UserSecrets`. `Docker Secrets` instead of `Secrets`

davidfowl commented on test/Microsoft.Extensions.Configuration.DockerSecrets.Test/DockerSecretTests.cs:
```diff
@@ -0,0 +1,247 @@
+{
+    public class DockerSecretTests
+    {
+
```
nit:space

HaoK commented:
Updated

HaoK commented on src/Microsoft.Extensions.Configuration.DockerSecrets/DockerSecretsConfigurationBuilderExtensions.cs:
```diff
@@ -0,0 +1,41 @@
+    /// <summary>
+    /// Extension methods for registering <see cref="DockerSecretsConfigurationProvider"/> with <see cref="IConfigurationBuilder"/>.
+    /// </summary>
+    public static class DockerSecretsConfigurationBuilderExtensions
```
@glennc can you take a look at the sugar methods and see if we need more (perhaps ones like optional)

HaoK replied:
Discussed with @glennc today, he doesn't feel strongly that we need optional but since I added it, we can keep it

HaoK commented:
We need to support nested keys in the filename for docker secrets, so we probably need to do something like what the Env provider is doing

Meir017 commented on src/Microsoft.Extensions.Configuration.DockerSecrets/DockerSecretsConfigurationProvider.cs:
```diff
@@ -0,0 +1,71 @@
+                }
+                else
+                {
+                    throw new FileNotFoundException("DockerSecrets directory doesn't exist and is not optional.");
```
maybe throw `DirectoryNotFoundException`

Meir017 commented on src/Microsoft.Extensions.Configuration.DockerSecrets/DockerSecretsConfigurationProvider.cs:
```diff
@@ -0,0 +1,71 @@
+            var secretsDir = Source.FileProvider.GetDirectoryContents("/");
+            if (!secretsDir.Exists && !Source.Optional)
+            {
+                throw new FileNotFoundException("DockerSecrets directory doesn't exist and is not optional.");
```
maybe throw `DirectoryNotFoundException`

HaoK commented:
@divega updated PR with what we discussed:
- Introduced `Add<TSource>(Action<TSource> configure) where TSource : IConfigurationSource, new()`, and added `AddXyz(Action<XyzSource> configure)` overloads to everything but AzureKeyVault which is weird because the provider/source are internal today unlike everything else...
- Using __ as key seperator in file names for docker secret similar to Env, and added a test for that

HaoK commented:
@glennc suggested we consider adding an additional `Func<string bool> FilterDelegate` which would be additional extensibility if they wanted to further customize what secrets to ignore.  We could default this to use our IgnorePrefix.  Thoughts @divega?

divega commented:
Re `FilterDelegate` I suggest we drop the word delegate from the name. Could be `Filter`, `FilterCondition` or similar. Also, it is not clear to me if the intent is to make it an inclusion or exclusion condition (filter is ambiguous). If it was exclusion then `IgnoreCondition` seems a good name.

I would also look at similar features in other Configuration providers. It seems they would all look different and that is not great.

HaoK commented:
Updated adding IgnoreCondition, with the default logic using starts with IgnoreFilter.  Setting to null turns of ignoring entirely.

HaoK commented:
Merging this as is for now after discussing with @glennc

HaoK commented:
2519ffc7fc1071befeae021551c6203126e117d3
